src/methods/widen.js

```diff
@@ -236,22 +236,35 @@ export class WidenImplementation {
 
           // For now, we only handle loop invariant properties
           //i.e. properties where the member expresssion does not involve any values written to inside the loop.
-          //todo: handle the case where key is an abstract value.
           let key = b.key;
-          if (typeof key === "string") {
-            pathNode = AbstractValue.createFromWidenedProperty(realm, rval, [b.object], ([o]) =>
-              t.memberExpression(o, t.identifier(key))
-            );
+          if (typeof key === "string" || !(key.mightNotBeString() && key.mightNotBeNumber())) {
+            if (typeof key === "string") {
+              pathNode = AbstractValue.createFromWidenedProperty(realm, rval, [b.object], ([o]) =>
+                t.memberExpression(o, t.identifier(key))
+              );
+            } else {
+              pathNode = AbstractValue.createFromWidenedProperty(realm, rval, [b.object, key], ([o, p]) => {
+                return t.memberExpression(o, p, true);
+              });
+            }
             // The value of the property at the start of the loop needs to be written to the property
             // before the loop commences, otherwise the memberExpression will result in an undefined value.
             let generator = realm.generator;
             invariant(generator !== undefined);
             let initVal = (b.descriptor && b.descriptor.value) || realm.intrinsics.empty;
             if (!(initVal instanceof Value)) throw new FatalError("todo: handle internal properties");
             if (!(initVal instanceof EmptyValue)) {
-              generator.emitVoidExpression(rval.types, rval.values, [b.object, initVal], ([o, v]) =>
-                t.assignmentExpression("=", t.memberExpression(o, t.identifier(key)), v)
-              );
+              if (key === "length" && b.object instanceof ArrayValue) {
+                // do nothing, the array length will already be initialized
+              } else if (typeof key === "string") {
+                generator.emitVoidExpression(rval.types, rval.values, [b.object, initVal], ([o, v]) =>
```

Comment:
This will make widenPropertyBindings() API having side-effect right? Every calling of widenPropertyBindings() will cause some entries being added into realm.generator. That may make the API harder to use(it can't be called multiple times). Not sure how to fix it though.

Reply:
This is a one time initialization of a "phi node" and will not be repeated if widenPropertyBindings is called again (as it usually is). Discovering the need for a phi node requires the fixed point computation, thus it cannot be reasonably separated from this code.